src/java.base/share/classes/jdk/internal/util/ReferencedKeyMap.java

```diff
@@ -336,6 +337,40 @@ public void removeStaleReferences() {
         }
     }
 
+    @SuppressWarnings("unchecked")
+    public void prepareForAOTCache() {
+        // We are running the AOT assembly phase. The JVM has a single Java thread, so
+        // we don't have any concurrent threads that may modify the map while we are
+        // iterating its keys.
+        //
+        // Also, the java.lang.ref.Reference$ReferenceHandler thread is not running,
+        // so even if the GC has put some of the keys on the pending ReferencePendingList,
+        // none of the keys would have been added to the stale queue yet.
+        assert CDS.isSingleThreadVM();
+
+        for (ReferenceKey<K> key : map.keySet()) {
+            Object referent = key.get();
+            if (referent == null) {
+                // We don't need this key anymore. Add to stale queue
+                ((Reference)key).enqueue();
```

Comment:
Is enqueue necessary here? Afaik this map only uses the queue to be alert of member reference being garbage collected and then remove stale elements. Since at this stage this map is no longer used, maybe `key.unused()` is sufficient?

Reply:
I'm not sure what you are proposing. But enqueue is used to ensure that inactive references are pruned from the data structure so that dumping only includes active references. This way, the JVM that loads these objects create these references as active and hence follow the usual life cycle that every other reference does. I want to avoid having a new "special" life cycles for dumped references.

Reply:
The map is still being used (and will be stored into the AOT cache). The key no longer has a referent, so we need to remove the key from the map. Adding the key to the stale queue and calling `removeStaleReferences()` will accomplish this. `key.unused()` does not remove the key from the map.

Reply:
Thanks. I am fine with the rest of the core library changes.